File: 0006-zigzag-conversion/0006-zigzag-conversion.js
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
var convert = function(s, numRows) {
  if (numRows === 1) return s;
  
  const ans = new Array(numRows).fill('');
  let [pos, direction] = [0, false];
  
  for (let i = 0; i < s.length; i++) {
    if (!(i % (numRows - 1))) direction = !direction;
    
    ans[pos] += s[i];
    direction ? pos++ : pos--;
  }
  
  return ans.join('');
};

File: 0006-zigzag-conversion/0006-zigzag-conversion.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1:
            return s
        ans = [''] * numRows
        pos, direction = 0, False
        
        for i, ch in enumerate(s):
            if i % (numRows - 1) == 0:
                direction = not direction
            ans[pos] += ch
            pos = pos + 1 if direction else pos - 1
        return ''.join(ans)
                